syntax = "proto3";

import "google/protobuf/timestamp.proto";

package posts;

service PostService {
    rpc CreatePost (CreatePostRequest) returns (PostResponse);
    rpc GetPost (GetPostRequest) returns (PostResponse);
    rpc UpdatePost (UpdatePostRequest) returns (PostResponse);
    rpc DeletePost (DeletePostRequest) returns (DeletePostResponse);
    rpc ListPosts (ListPostsRequest) returns (ListPostsResponse);
}

message Post {
    string id = 1;
    string title = 2;
    string description = 3;
    string creator_id = 4;
    google.protobuf.Timestamp created_at = 5;
    google.protobuf.Timestamp updated_at = 6;
    bool is_private = 7;
    repeated string tags = 8;
}

message CreatePostRequest {
    string title = 1;
    string description = 2;
    string creator_id = 3;
    bool is_private = 4;
    repeated string tags = 5;
}

message GetPostRequest {
    string id = 1;
    string user_id = 2;
}

message UpdatePostRequest {
    string id = 1;
    string title = 2;
    string description = 3;
    string updater_id = 4; 
    bool is_private = 5;
    repeated string tags = 6;
}

message DeletePostRequest {
    string id = 1;
    string deleter_id = 2;
}

message DeletePostResponse {
    bool success = 1;
}

message ListPostsRequest {
    int32 page = 1;
    int32 per_page = 2;
    string user_id = 3;
}

message ListPostsResponse {
    repeated Post posts = 1;
    int32 total = 2;
    int32 page = 3;
    int32 per_page = 4;
}

message PostResponse {
    Post post = 1;
}